File: src/dexter/frontend/clang/src/ir/bool/FieldExpr.cpp
/*
 * FieldExpr.cpp
 *
 *  Created on: Mar 24, 2018
 *      Author: maazsaf
 */

#include "ir/bool/FieldExpr.h"

Dexter::FieldExpr::FieldExpr (Dexter::Expr * target, std::string field) : Dexter::Expr(NULL)
{
  JNIEnv * e = Util::env;
  jclass c = e->FindClass("dexter/ir/bool/FieldExpr");
  jmethodID m = e->GetMethodID(c, "<init>", "(Ldexter/ir/Expr;Ljava/lang/String;)V");
  jstring f = e->NewStringUTF(field.c_str()) ;
  this->mObj = e->NewObject(c, m, target->obj(), f);
}

File: src/dexter/ir/bool/FieldExpr.java
package dexter.ir.bool;

import dexter.ir.Expr;
import dexter.ir.Visitor;

/**
 * Created by Maaz Ahmad on 6/25/19.
 */
public class FieldExpr extends Expr
{
  protected Expr target;
  protected String field;

  public FieldExpr(Expr target, String field)
  {
    this.target = target;
    this.field = field;
  }

  public Expr target () { return target; }
  public String field () { return field; }

  @Override
  public <T> T accept(Visitor<T> v)
  {
    return v.visit(this);
  }

  @Override
  public boolean equals (Object o)
  {
    if (o instanceof FieldExpr)
    {
      FieldExpr other = (FieldExpr)o;
      return other.target.equals(this.target) && other.field.equals(this.field);
    }
    else
      return false;
  }
}